demos/mqtt/mqtt_demo_connection_sharing/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,308 @@
+        /* Any character mismatch other than '+' or '#' means the topic
+         * name does not match the topic filter. */
+        *pMatch = false;
+        status = true;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool topicFilterMatch( const char * pTopicName,
+                              uint16_t topicNameLength,
+                              const char * pTopicFilter,
+                              uint16_t topicFilterLength )
+{
+    bool status = false, matchFound = false;
+    uint16_t nameIndex = 0, filterIndex = 0;
+
+    while( ( nameIndex < topicNameLength ) && ( filterIndex < topicFilterLength ) )
+    {
+        /* Check if the character in the topic name matches the corresponding
+         * character in the topic filter string. */
+        if( pTopicName[ nameIndex ] == pTopicFilter[ filterIndex ] )
+        {
+            /* Handle special corner cases regarding wildcards at the end of
+             * topic filters, as documented by the MQTT protocol spec. */
+            matchFound = matchEndWildcards( pTopicFilter,
+                                            topicNameLength,
+                                            topicFilterLength,
+                                            nameIndex,
+                                            filterIndex,
+                                            &status );
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            matchFound = matchWildcards( pTopicFilter,
+                                         pTopicName,
+                                         topicNameLength,
+                                         filterIndex,
+                                         &nameIndex,
+                                         &status );
+        }
+
+        if( matchFound == true )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( status == false )
+    {
+        /* If the end of both strings has been reached, they match. */
+        status = ( ( nameIndex == topicNameLength ) && ( filterIndex == topicFilterLength ) );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool topicMatch( const char * pTopicName,
+                        const uint16_t topicNameLength,
+                        const char * pTopicFilter,
+                        const uint16_t topicFilterLength,
+                        bool requiresExactMatch )
+{
+    bool status = false;
+
+    /* Check for an exact match. */
+    if( topicNameLength == topicFilterLength )
+    {
+        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 );
+    }
+
+    /* If  an exact match is required, return the result of the comparison above.
+     * Otherwise, attempt to match with MQTT wildcards in topic filters. */
+    if( requiresExactMatch == false )
```

Comment:
Would we ever want to set this to true? I see you don't and that makes sense to me, so I think this function could just be removed and `topicFilterMatch` used in its place

Reply:
I took the topic matching logic from the `v4_beta` branch [here](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/libraries/standard/mqtt/src/iot_mqtt_subscription.c#L316-L321). 
Looking at the call-site of this function in the `v4_beta` implementation, it seems like this parameter was added for a different use-case (of updating registered callback against the same topic for which subscription already exists).

I will remove the `requiresExactMatch` logic